### DISCUSSION TITLE
How do achieve variations feature from MJ/A11?

### DISCUSSION THREAD

**Discussion by morphles** [Comfy-Org/ComfyUI#534] [Q&A]
I suspect I need to sue advanced KSampler, with small noise addition? Overall, I suspect I could do some crazy shit with KSampler advanced, but I have not yet touched it, as I have not seen any docs about it. Is there any?

**Comment by omar92** [comment]
i would suggest that you use 2 advanced ksamplers , one do like 10 steps then the other do the remaining 10 steps ,, but  change the seed on the secound one every time you run so you get variants

**Reply by catboxanon** [reply]
This isn't really the same thing. A1111 performs a slerp for two sets of noise.

https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/22bcc7be428c94e9408f589966c2040187245d81/modules/processing.py#L404

**Reply by comfyanonymous** [reply]
Is this something that's actually useful? Doing a lerp between two noises is not going to work very well but is slerp actually the best way to interpolate between noises for stable diffusion?

**Reply by catboxanon** [reply]
I've personally used it a lot so I would say it's useful, and is likely more in line with what @morphles is asking about. I can't comment on it being the best method for SD. According to [the discussion](https://discuss.pytorch.org/t/help-regarding-slerp-function-for-generative-model-sampling/32475/3) that's linked to in the code, it was used by NVIDIA for GAN 2.0, which in [the paper](https://arxiv.org/abs/1812.04948) they claim in section 4.1 is "the most appropriate way of interpolating in our normalized input latent space".

**Comment by omar92** [comment]
so in conclusion  how to do that variants in comfy ?

**Comment by BlenderNeko** [comment]
If other nodes could provide the sampler nodes with their own input noise pattern this would be one of the things that a set of custom nodes could provide.

@comfyanonymous What do you think would be the best way for the samplers to receive such an input? An optional parameter? Under a specific key inside the latent dict? Or do you have other ideas/plans?

**Reply by comfyanonymous** [reply]
What kind of noise input pattern are you thinking about feeding the samplers?

**Reply by BlenderNeko** [reply]
any kind of noise that simply replaces [this](https://github.com/comfyanonymous/ComfyUI/blob/master/nodes.py#L749) variable when supplied. Creating variants can then be offloaded to some custom nodes, there'll be less hacky ways for users to use GPU generated noise, a potential custom node that does the reverse noise stuff from [here](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/scripts/img2imgalt.py) could also make use of it.

**Reply by comfyanonymous** [reply]
If I expose a way to get the sigmas would that solve the problem since KSamplerAdvanced supports not adding noise to the latent already?

**Reply by BlenderNeko** [reply]
I think that should also work.. I was a bit hesitant in suggesting that as it seemed more complex to me, but maybe just some refactoring in the samplers file would be sufficient.

Feels like that would probably be cleaner in the long run? yeah lets do that.

**Reply by BlenderNeko** [reply]
i'll probably have PR for this soonish

**Comment by BlenderNeko** [comment]
Alright you can now use the nodes in this [noise repo](https://github.com/BlenderNeko/ComfyUI_Noise) to create variations like in A1111. It's not exactly the same as A1111 though, cuz A1111 borked the math involved.

**Reply by catboxanon** [reply]
Awesome, cheers!

**Reply by Ai-Enrapture** [reply]
How do I use this to get a unique seed that is replicable but gives variation like A1111?

**Reply by Ai-Enrapture** [reply]
Also the seed doesn't want to change, but instead seems locked or resets to the previous seed when rendering. It's a bug right?

**Comment by chrisgoringe** [comment]
I wanted something similar, so I wrote this pair of really simple custom nodes - https://github.com/chrisgoringe/cg-noise - which hijack the noise generation and replace it...
